types/wlr_output.c

```diff
@@ -642,6 +657,74 @@ static void output_pending_resolution(struct wlr_output *output, int *width,
 	}
 }
 
+static bool output_attach_empty_buffer(struct wlr_output *output) {
+	assert(!(output->pending.committed & WLR_OUTPUT_STATE_BUFFER));
+
+	if (!wlr_output_attach_render(output, NULL)) {
+		return false;
+	}
+
+	int width, height;
+	output_pending_resolution(output, &width, &height);
+
+	struct wlr_renderer *renderer = wlr_backend_get_renderer(output->backend);
+	wlr_renderer_begin(renderer, width, height);
+	wlr_renderer_clear(renderer, (float[]){0, 0, 0, 0});
+	wlr_renderer_end(renderer);
+
+	return true;
+}
+
+static bool output_ensure_buffer(struct wlr_output *output) {
+	// If we're lighting up an output or changing its mode, make sure to
+	// provide a new buffer
+	bool needs_new_buffer = false;
+	if ((output->pending.committed & WLR_OUTPUT_STATE_ENABLED) &&
+			output->pending.enabled) {
+		needs_new_buffer = true;
+	}
+	if (output->pending.committed & WLR_OUTPUT_STATE_MODE) {
+		needs_new_buffer = true;
+	}
+	if (!needs_new_buffer ||
+			(output->pending.committed & WLR_OUTPUT_STATE_BUFFER)) {
+		return true;
+	}
+
+	wlr_log(WLR_DEBUG, "Attaching empty buffer to output for modeset");
+
+	if (!output_attach_empty_buffer(output)) {
+		output_clear_back_buffer(output);
+		return false;
+	}
+	if (!output->impl->test || output->impl->test(output)) {
+		return true;
+	}
+
+	output_clear_back_buffer(output);
+
+	if (output->swapchain->format->len == 0) {
+		return false;
+	}
+
+	// The test failed for a buffer which has modifiers, try disabling
+	// modifiers to see if that makes a difference.
+	wlr_log(WLR_DEBUG, "Output modeset test failed, retrying without modifiers");
+
+	if (!output_create_swapchain(output, false)) {
```

Comment:
So we set up a modifier-less swapchain here, but the code below will in some cases try to allocate a modifier...full swapchain again: `output_attach_empty_buffer` below calls `wlr_output_attach_render`, which in case of lacking `attach_render` implementation will call `output_attach_back_buffer` and ultimately attempt to allocate a swapchain with modifiers.

It would be better if we could avoid the split responsibility for swapchain setup.

Reply:
This shouldn't happen, since the `output_create_swapchain` call before ensures the swapchain is already allocated, so `output_attach_empty_buffer` won't do it again. I agree it's not ideal, but I don't have great ideas to simplify it.

Reply:
Ah, you're right, it is correct although a bit subtle. Note to self: If you `output_create_swapchain` with `allow_modifiers = false`, then only modifier-less swapchains will be reused, while if called with `allow_modifiers = true`, then any swapchain will be reused rather than giving you a modifier-full one.

It wouldn't hurt with a comment at output_create_swapchain to describe this behavior, as it's definitely more subtle than the previous swapchain reuse logic.

Reply:
Yes, that's a good point. Added a comment.